### PR TITLE
[DOC] add runtime abstraction layer doc

### PR DESCRIPTION
**Pull request by wyzero**

### CODE REVIEW

File: docs/developers/runtime_abstraction_layer.md [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+
+## 2.1 From the Perspective of Compiler Side
+
+In the compiler side, RAL has a set of transformation passes to make sure the
+compiled binary is suitable for the RAL runtime.
+
+### 2.1.1 Context Injection
+
+As mentioned above，we prefer to simplify the design of the compiler side and
+let it foucus on its core optimization logic. Thus, RAL provides a context
```
Comment (yangjunpro) [inline docs/developers/runtime_abstraction_layer.md:33]: Nit."focus"

File: docs/developers/runtime_abstraction_layer.md [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+pointer in llvm IR.
+
+### 2.1.2 Inputs/Outputs Binding
+
+For the entry function, RAL also rewrites its inputs and outputs to make it
+suitable to interact with different host environments. To be concrete, all the
+original inputs and outputs of the entry function are received from and sent to
+RAL through a sequence of RAL API calls correspondingly. The motivation behind
+this is to hide the implementation details of memref structure to make the ABI
+clean and stable.  Below is an example IR before and after the transformation.
```
Comment (yangjunpro) [inline docs/developers/runtime_abstraction_layer.md:52]: Nit."an example IR"==>"a sample IR snippet"

File: docs/developers/runtime_abstraction_layer.md [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+  use(%arg0, %arg1, %ret, ...)
+  disc_ral.send_output(%ctx, 0, %ret) // send the first outpt
+}
+```
+
+As shown in the above example, the inputs (similiar to the outputs) are received
+from the ral context one by one. We could choose to use one api call to receive
+all inputs at the same time, which may also have lower runtime overhead in terms
+of the number of ral api calls. However, the current design lets the compiler
+have the opportunity the adjust the placement of each recv_input and send_output
```
Comment (yangjunpro) [inline docs/developers/runtime_abstraction_layer.md:80]: Nit."have the opportunity the adjust"==>"have the opportunity to adjust".

File: docs/developers/runtime_abstraction_layer.md [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+api to achieve partial execution of a compiled binary before all its inputs are
```
Comment (yangjunpro) [inline docs/developers/runtime_abstraction_layer.md:81]: Nit. Better to give a small example here. Partial execution is a nice feature, but maybe a little bit against normal intuition.

File: docs/developers/runtime_abstraction_layer.md [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+ready and sending some of the outputs to the context before all its outputs are
+ready.
+
+# 2.1.3 Uniformed type-erased external function call ABI
+To make the compiled binary having stable and clean ABI, RAL provides a pass to
```
Comment (yangjunpro) [inline docs/developers/runtime_abstraction_layer.md:86]: Nit."make the compiled binary having"==>"make the compiled binary have".

File: docs/developers/runtime_abstraction_layer.md [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+rewrite all ral function calls to a  Uniformed type-erased style.
```
Comment (yangjunpro) [inline docs/developers/runtime_abstraction_layer.md:87]: Nit.“Uniformed"=>"uniformed"?

File: docs/developers/runtime_abstraction_layer.md [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+
+A RAL function is an ordinary c++ function which satisfies the following
+requirements.
+* The first argument of the function should be a pointer to RAL context.
+* RAL supports type bool, int8/16/32/64, half/float/double, MemRef struct and
+  ordinary pointer.
+* The types of remaining arguments should be supported by RAL.
+* The returning type should be supported by RAL.
```
Comment (yangjunpro) [inline docs/developers/runtime_abstraction_layer.md:95]: Nit."returning type"==>"returned type"?

File: docs/developers/runtime_abstraction_layer.md [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+
+For a given RAL function, it has a equivalent interface, namely a type-erased c
```
Comment (yangjunpro) [inline docs/developers/runtime_abstraction_layer.md:97]: Nit."a equivalent"==>"an equivalent"

File: docs/developers/runtime_abstraction_layer.md [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+language interface. The main purpose of this kind of interface is used for cross
+language binding and linking.
+
+Below is an example to demonstrate the c++ format and correspondding uniform
+type-erased format.
+
+```
+// A RAL API to do GEMM operation, c++ format
+void gemm(RalContext*, gpu_stream_handle*, MemRef<float, 2> lhs, MemRef<float, 2> rhs, MemRef<float, 2> out, bool lhs_transposed, bool, rhs_transposed);
```
Comment (yangjunpro) [inline docs/developers/runtime_abstraction_layer.md:106]: Nit."bool, rhs_transposed"=>"bool rhs_transposed".

File: docs/developers/runtime_abstraction_layer.md [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+//
+//     inputs_encode = type_separator.join([type_encoding for type in
+//     input_types])
+//
+//     outputs_encode = type_separator.join([type_encoding for type
+//     in output_types])
+//
+//     type_separator = '_'
+
+// The dispatcher will dispatch to concret RAL function implementation according to the api_name
```
Comment (yangjunpro) [inline docs/developers/runtime_abstraction_layer.md:133]: Nit."concret"=>"concrete”

File: docs/developers/runtime_abstraction_layer.md [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+void ral_api_call(void* ctx, const char* api_name, void** args);
+```
+
+Furthermore, we also provide a registry mechanism to automatically convert a RAL
```
Comment (yangjunpro) [inline docs/developers/runtime_abstraction_layer.md:137]: Nit."registry"==>"registration"?

File: docs/developers/runtime_abstraction_layer.md [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+function from its c++ format to its uniform format. Below is an example.
+
+```
+// A RAL API to do GEMM operation, c++ format
+template <typename T>
+void gemm(RalContext*, gpu_stream_handle*, MemRef<T, 2> lhs, MemRef<T, 2> rhs, MemRef<T, 2> out, bool lhs_transposed, bool, rhs_transposed);
```
Comment (yangjunpro) [inline docs/developers/runtime_abstraction_layer.md:143]: Nit."bool, rhs_transposed" :)

File: docs/developers/runtime_abstraction_layer.md [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+can use the c++ level api directly.
+
+## 2.2，From the Perspective of Runtime Side
+
+### 2.2.1 Overview of the Runtime Side of RAL
+
+The runtime side of the RAL can be divided into three major parts as shown in
+the following ![diagram](./RAL.png).
+
+* Interacting with the host environments (e.g. Tensorflow, PyTorch). This part
```
Comment (yangjunpro) [inline docs/developers/runtime_abstraction_layer.md:170]: NIt.Better to use "TensorFlow" to be consistent with "PyTorch".

File: docs/developers/runtime_abstraction_layer.md [hunk 14]
```diff
@@ -0,0 +1,208 @@
...
+  apis. The driver ususlly includes memory management, task launching and
+  synchronizaiton. It's worth noting that we need different driver
+  implementations for the same device type in different host environments. One
+  reason is that different host environment usually have different device
+  abstraction layer, leading to different api style and semantics. Another
+  reason is that using host environment native device resource access style
+  could also minimize the overhead when crossing host enviroment and RAL
+  boundary.
+* Providing an implementation of the custom kernel. Examples are vendor library
+  kernel, RNG, TopK or sort op. These kernels are ususlly not codegen by
```
Comment (yangjunpro) [inline docs/developers/runtime_abstraction_layer.md:184]: Nit.“ususlly"==>"usually".

File: docs/developers/runtime_abstraction_layer.md [hunk 15]
```diff
@@ -0,0 +1,208 @@
...
+  BladeDISC and rely on some mannual crafted implementations. It worth noting
```
Comment (yangjunpro) [inline docs/developers/runtime_abstraction_layer.md:185]: Nit."mannual"==>"manual"

File: docs/developers/runtime_abstraction_layer.md [hunk 16]
```diff
@@ -0,0 +1,208 @@
...
+  that the custom kernel itself uses the ral driver api to access device
+  resoruces, and thus can be shared among different host environments since the
```
Comment (yangjunpro) [inline docs/developers/runtime_abstraction_layer.md:187]: Nit."resoruces"=>”resources".

File: docs/developers/runtime_abstraction_layer.md [hunk 17]
```diff
@@ -0,0 +1,208 @@
...
+  ral dirver has already hided the differences among different host
+  environments.
+
+### 2.2.2 RAL Context
+
+We provide different RAL Context implementation for different target host
+environments. The context implementation consists of the above three major
+parts. Currently we have tf context implementation, which is used to be
+integrated with tensorflow (both cpu & gpu), and base context implmentation,
+which is used to be integrated with Pytorch and sometimes the standalone binary.
```
Comment (yangjunpro) [inline docs/developers/runtime_abstraction_layer.md:197]: Nit. For literal "Pytorch", I think either "PyTorch" or "Pytorch" should be fine, assuming that we use it in a consistent approach.

File: docs/developers/runtime_abstraction_layer.md [hunk 18]
```diff
@@ -0,0 +1,208 @@
...
+
+### 2.2.3 RAL Execution Context
+
+For a loaded compiled binary, we have one corresponding RAL context to manage
+the stateful resources and some bookkeeping stuff.  The context lifetime is same
```
Comment (yangjunpro) [inline docs/developers/runtime_abstraction_layer.md:202]: Nit."same"=>"the same"

File: docs/developers/runtime_abstraction_layer.md [hunk 19]
```diff
@@ -0,0 +1,209 @@
...
+to hide all stateful operations behind the context. The context is passed as a
+parameter to the entry function of the compilation module and all RAL APIs
+should always use the context as their first argument. RAL uses a context
+injection pass to ensure this property. The pass rewrites the entry function and
+all related functions to make sure their first argument is the context. To be
+concrete, we create a custom dialect disc_ral using MLIR infra to model the RAL
+runtime behavior. Inside the dialect, we define a custom type
+disc_ral.RalExecutionContextType to represent the context type in the compiler
+side. Under the hood, the disc_ral.RalExecutionContextType will be lowered to a
+pointer in LLVM IR.
```
Comment (linearhit) [inline docs/developers/runtime_abstraction_layer.md:43]: disc_ral.RalExecutionContextType will be lowered to a pointer type in LLVM IR

File: docs/developers/runtime_abstraction_layer.md [hunk 20]
```diff
@@ -0,0 +1,209 @@
...
+
+### 2.1.2 Inputs/Outputs Binding
+
+For the entry function, RAL also rewrites its inputs and outputs to make it
+suitable to interact with different host environments. To be concrete, all the
+original inputs and outputs of the entry function are received from and sent to
+RAL through a sequence of RAL API calls correspondingly. The motivation behind
+this is to hide the implementation details of the MemRef structure to make the
+ABI clean and stable.  Below is the example IR before and after the
```
Comment (linearhit) [inline docs/developers/runtime_abstraction_layer.md:52]: One more space before ’Below'

File: docs/developers/runtime_abstraction_layer.md [hunk 21]
```diff
@@ -0,0 +1,209 @@
...
+  APIs. The driver usually includes memory management, task launching, and
+  synchronization. It's worth noting that we need different driver
+  implementations for the same device type in different host environments. One
+  rationale is that different host environments usually have different device
+  abstraction layers, leading to different API styles and semantics. Another
+  reason is that using the host environment native device resource access style
+  could also minimize the overhead when crossing the host environment and RAL
+  boundary.
+* Providing an implementation of the custom kernel. Examples are vendor library
+  kernel, RNG, TopK, or sort op. These kernels are usually not codegen by
```
Comment (linearhit) [inline docs/developers/runtime_abstraction_layer.md:185]: codegen -> code generated, or just 'generated'

**Review by linearhit — APPROVED** [review]
LGTM

**Pull request merged**